packages/semver/src/builders/version/plugin-adapter.ts

```diff
@@ -0,0 +1,116 @@
+import { resolve } from 'path';
+import { concat, from, Observable } from 'rxjs';
+import { switchMap } from 'rxjs/operators';
+
+import { PluginOptions, PluginType, SemverPlugin } from './plugin';
+import { getOutputPath, getProjectRoot } from './utils/workspace';
+import { CommonVersionOptions } from './version';
+
+export interface RawSemanticReleasePlugin {
+  addChannel?(...args: SemanticReleasePluginOptions): Promise<unknown>;
+  publish?(...args: SemanticReleasePluginOptions): Promise<unknown>;
+}
+
+export interface SemanticReleaseContext {
+  cwd: string;
+  env: NodeJS.ProcessEnv;
+  stdout: NodeJS.WriteStream;
+  stderr: NodeJS.WriteStream;
+  nextRelease: { version: string; channel?: string };
+  logger: { log(msg: string): void };
+}
+
+export type SemanticReleasePluginOptions = [
+  pluginOptions: { npmPublish: boolean; pkgRoot: string },
+  context: SemanticReleaseContext
+];
+
+export class SemanticReleasePlugin implements SemverPlugin {
+  name: string;
+
+  type: PluginType = '@semantic-release';
+
+  private _plugin: RawSemanticReleasePlugin;
+
+  constructor({
+    name,
+    plugin,
+  }: {
+    name: string;
+    plugin: RawSemanticReleasePlugin;
+  }) {
+    this.name = name;
+    this._plugin = plugin;
+  }
+
+  publish(
+    _: PluginOptions,
+    options: CommonVersionOptions,
+    context: BuilderContext
+  ): Observable<unknown> {
+    return from(_createOptions(options, context)).pipe(
+      switchMap((options) =>
+        concat(
+          this._plugin.addChannel(...options),
+          this._plugin.publish(...options)
+        )
+      )
+    );
+  }
+}
+
+export async function _createOptions(
+  options: CommonVersionOptions,
+  context: BuilderContext
+): Promise<SemanticReleasePluginOptions> {
+  const pkgRoot = await getOutputPath(context).toPromise();
+  const projectRoot = await getProjectRoot(context).toPromise();
+
+  return [
+    {
+      npmPublish: options.dryRun === false,
+      pkgRoot,
+    },
+    {
+      cwd: projectRoot,
+      env: process.env,
+      stdout: process.stdout,
+      stderr: process.stderr,
+      nextRelease: { version: options.newVersion }, // @todo handle channel option
+      logger: { log: context.logger.info },
+    },
+  ];
+}
+
+
+export class PluginAdapter {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static adapt({ name, plugin }: { name: string; plugin: any }): SemverPlugin {
+    switch (true) {
+      case _isSemanticPlugin(plugin):
+        return new SemanticReleasePlugin({ name, plugin });
+
+      case _isSemverPlugin(plugin):
+        return plugin;
+
+      default:
+        throw new Error(`Plugin not supported`);
+    }
+  }
+}
+
+export function _isSemanticPlugin(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  plugin: any
+): plugin is RawSemanticReleasePlugin {
+  return (
+    typeof plugin.publish === 'function' &&
+    typeof plugin.verifyCondition === 'function'
```

Comment:
I think that all functions are optional. Apparently `analyzeCommits` is required but I think that what the documentation means is that at least one of the plugins should have the function.
TL;DR: There could be some plugins that have neither `publish` nor `verifyCondition`. https://github.com/semantic-release/semantic-release/blob/master/docs/usage/plugins.md

Reply:
Yep I was not sure about it, there's no easy way to do duck typing here. Maybe we can check if one hook exist at least?

Reply:
this sounds perfect 👌